ssl/quic/quic_impl.c

```diff
@@ -4324,10 +4324,119 @@ SSL *ossl_quic_new_listener_from(SSL *ssl, uint64_t flags)
+
+    if (!expect_quic_listener(ssl, &ctx))
+        return NULL;
+
+    if (!SSL_up_ref(&ctx.ql->obj.ssl))
+        return NULL;
+
+    qctx_lock(&ctx);
+
+    ql = ctx.ql;
+
+    if ((qc = OPENSSL_zalloc(sizeof(*qc))) == NULL) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_CRYPTO_LIB, NULL);
+        SSL_free(&ql->obj.ssl);
+        goto err;
+    }
+
+    /*
+     * XXX setting listener here is needed so `qc_cleanup()` does
+     * a right thing. There is `TOOD` comment in `qc_cleanup()`, which
+     * makes me believe line below is correct for now.
+     * Without this might be destroying port prematurely.
+     *
+     * Perhaps we should be grabbing a references to engine and port here.
+     */
+    qc->listener = ql;
+    qc->engine = ql->engine;
+    qc->port = ql->port;
+/* create channel */
+#if defined(OPENSSL_THREADS)
+    /* this is the engine mutex */
+    qc->mutex = ql->mutex;
+#endif
+#if !defined(OPENSSL_NO_QUIC_THREAD_ASSIST)
+    qc->is_thread_assisted
+    = ((ql->obj.domain_flags & SSL_DOMAIN_FLAG_THREAD_ASSISTED) != 0);
+#endif
+
+    /* Create the handshake layer. */
+    qc->tls = ossl_ssl_connection_new_int(ql->obj.ssl.ctx, NULL, TLS_method());
+    if (qc->tls == NULL || (sc = SSL_CONNECTION_FROM_SSL(qc->tls)) == NULL) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_INTERNAL_ERROR, NULL);
+        goto err;
+    }
+    sc->s3.flags |= TLS1_FLAGS_QUIC;
+
+    qc->default_ssl_options = OSSL_QUIC_PERMITTED_OPTIONS;
+    qc->last_error = SSL_ERROR_NONE;
+
+    /* this is QCSO, we don't expect to accept connections */
+    qc->ch = ossl_quic_port_create_outgoing(qc->port, qc->tls);
+    if (qc->ch == NULL)
+        goto err;
+
+    ossl_quic_channel_set_msg_callback(qc->ch, ql->obj.ssl.ctx->msg_callback, &qc->obj.ssl);
+    ossl_quic_channel_set_msg_callback_arg(qc->ch, ql->obj.ssl.ctx->msg_callback_arg);
+
+    /*
+     * we deliberately pass NULL for engine and port, because we don't want to
+     * to turn QCSO we create here into an event leader, nor port leader,
+     * because the listener (`ssl`) took both those roles already. The question
+     * is what happens with listener gets destroyed, who will be the leader then?
+     * I still need to convince myself there is a plan for this.
+     */
+    if (!ossl_quic_obj_init(&qc->obj, ql->obj.ssl.ctx,
+                            SSL_TYPE_QUIC_CONNECTION,
+                            &ql->obj.ssl, NULL, NULL)) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_INTERNAL_ERROR, NULL);
+        SSL_free(&ql->obj.ssl);
+        goto err;
+    }
+
+    /* Initialise libssl APL-related state. */
+    qc->default_stream_mode = SSL_DEFAULT_STREAM_MODE_AUTO_BIDI;
+    qc->default_ssl_mode = qc->obj.ssl.ctx->mode;
+    qc->default_ssl_options = qc->obj.ssl.ctx->options & OSSL_QUIC_PERMITTED_OPTIONS;
+    qc->incoming_stream_policy = SSL_INCOMING_STREAM_POLICY_AUTO;
+    qc->last_error = SSL_ERROR_NONE;
+    /* qc->obj.ssl.method = OSSL_QUIC_client_method(); */
```

Comment:
This deserves some explanation.

Reply:
sorry the line 4424 is left behind. I was curious to see what happens if I change SSL method from _server_method we currently use here to _client_method() nothing happened. At least I could not spot any difference. Will drop the line.